ui: use uPlot lib for line graphs on metrics page

dhartunian opened the pull request:
Previously, we were using the nvd3 library to render line graphs. This
library is deprecated and relied on svg for rendering visuals. For
clusters past a certain size (~30 nodes) performance on the metrics
dashboards would become unusably slow due to the large amount of svg on
the page. In addition, it looks like nvd3 would render a lot of
unnecessary svg along with the lines to render points on hover which was
difficult to circumvent.

This PR replaces the `LineGraph` component with one that uses the uPlot
library (https://github.com/leeoniya/uPlot). This library uses the
canvas API to render and has been engineered with performance in mind
for larger datasets. The implementation in this PR can render a
60-cluster Hardward metrics page quite quickly while retaining a
guideline on hover with legend displaying the values at the guideline.

Here's a list of important features implemented with uPlot in this commit
- Plotting graphs with correctly formatted x-axis
- Hovering legend on mouseover that mimics what we had before (legend
redesign is pending future work)
- Aggregation of multiple "result" time series from CRDB into a single
set of y values aligned against a single time series (this is how uPlot
requires its data to be formatted, we previously had separate time axes
per plot series)
- Interactive selection of zoom area on uPlot graph that triggers a zoom
on the entire dashboard
- uPlot uses same color palette as nvd3 implementation

Removed features from nvd3 implementation:
- Synced guideline on all graphs in dashboard

Resolves #62463.

Release note (ui change): use a new library for line graphs that renders
metrics more efficiently. Customers with large clusters can now load and
interact with metrics much faster than before. (TODO: what features are
different with this new library?)

dhartunian committed:
ui: use uPlot lib for line graphs on metrics page

Previously, we were using the nvd3 library to render line graphs. This
library is deprecated and relied on svg for rendering visuals. For
clusters past a certain size (~30 nodes) performance on the metrics
dashboards would become unusably slow due to the large amount of svg on
the page. In addition, it looks like nvd3 would render a lot of
unnecessary svg along with the lines to render points on hover which was
difficult to circumvent.

This PR replaces the `LineGraph` component with one that uses the uPlot
library (https://github.com/leeoniya/uPlot). This library uses the
canvas API to render and has been engineered with performance in mind
for larger datasets. The implementation in this PR can render a
60-cluster Hardward metrics page quite quickly while retaining a
guideline on hover with legend displaying the values at the guideline.

Here's a list of important features implemented with uPlot in this commit
- Plotting graphs with correctly formatted x-axis
- Hovering legend on mouseover that mimics what we had before (legend
redesign is pending future work)
- Aggregation of multiple "result" time series from CRDB into a single
set of y values aligned against a single time series (this is how uPlot
requires its data to be formatted, we previously had separate time axes
per plot series)
- Interactive selection of zoom area on uPlot graph that triggers a zoom
on the entire dashboard
- uPlot uses same color palette as nvd3 implementation

Removed features from nvd3 implementation:
- Synced guideline on all graphs in dashboard

Resolves #62463.

Release note (ui change): use a new library for line graphs that renders
metrics more efficiently. Customers with large clusters can now load and
interact with metrics much faster than before. (TODO: what features are
different with this new library?)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63087)
<!-- Reviewable:end -->

leeoniya commented:
btw, you can set `ms: 1` and pass milliseconds in the data.

https://github.com/leeoniya/uPlot/blob/f8f1b01aa68c8869c03b5e205e8fc425b1912128/dist/uPlot.d.ts#L324-L325

dhartunian commented:
> btw, you can set `ms: 1` and pass milliseconds in the data.
> 
> https://github.com/leeoniya/uPlot/blob/f8f1b01aa68c8869c03b5e205e8fc425b1912128/dist/uPlot.d.ts#L324-L325

Wonderful, thanks for the tip! Was going to ping you via #45171 once I was a bit further along but you're already on it 😄

nathanstilwell reviewed (commented):
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63087)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, @koorosh, and @nkodali)

___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgNJvu2U5s5xl16lx_:-MXgNJvu2U5s5xl16lxa:bdiuj1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L308)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>   const xValues: number[] = [];
>   data.results[0].datapoints.forEach((datapoint) => {
>     xValues.push(NanoToSeconds(datapoint.timestamp_nanos.toNumber()));
>   });
> ```
</details>


This looks a lot like a `.map`.

```suggestion
const xValues: number[] = data.results[0].datapoints.map(d => 
    NanoToSeconds(d.timestamp_nanos.toNumber()))
```
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgNwtkEB2S3lv8vGOQ:-MXgNwtkEB2S3lv8vGOR:b-43sjfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L313)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>   const yValues: (number | null)[][] = [];
>   data.results.forEach((result) => {
>     yValues.push(result.datapoints.map((d) => d.value));
>   });
> ```
</details>


... maybe for this one a `.reduce`?

```js
  const yValues: (number | null)[][] = 
  data.results.reduce(
    (values, result) => [...values, result.datapoints.map((d) => d.value)],
    [],
  );
```
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgU0YN1s8SLBzkbzY5:-MXgU0YN1s8SLBzkbzY6:by29p1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L371)):*
> ```XML
>   componentDidUpdate(prevProps: Readonly<LineGraphProps>) {
>     if (
>       // data was missing last time or we already have a `u`
> ```

Thank you for the comment :pray: 
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 393 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgUKHb3YRgadABSrn8:-MXgUKHb3YRgadABSrn9:b-t9jlpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L393)):*
> ```XML
>       this.u = new uPlot(options, uPlotData, this.el);
>     }
>     if (this.u && prevProps.data != this.props.data) {
> ```

A couple comments. 

1. This is JavaScript so, "strict equality" please. :smile: (`!==` vs `!=`)
2. Isn't `data` an object? Are you expecting a _deep equals_ here? Or is this just to check that the props changed (in which case this works because it's a _different object_). **






** I haven't finished my coffee yet this morning, so I'm betting this is just a props change check. If so feel free to ignore. 
___
*[pkg/ui/src/views/cluster/util/graphs.ts, line 595 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgWVdDDOHZ_SUd30Vg:-MXgWVdDDOHZ_SUd30Vh:b-r82nyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/util/graphs.ts#L595)):*
> ```XML
>       //
>       // Note: Any custom range setting here will need to get
>       // auto-updated when the timescale or data changes.a
> ```

just pointing out a _dangling_ `a` on the end of this sentence


<!-- Sent from Reviewable.io -->

koorosh reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63087)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, and @nkodali)

___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY4aPwb6-5_odD448nR:-MY4aPwb6-5_odD448nS:bl2ywoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L309)):*
> ```XML
>   //insert x values
>   const xValues: number[] = [];
>   data.results[0].datapoints.forEach((datapoint) => {
> ```

@dhartunian , current data points transformation function relies on the shape of the first series and may work in case all series have the same length.
But it looks like series can have different length that might cause incorrect points mapping.
As an example, if one node is dead, Hardware metrics don't include any data points for such node
![Screen Shot 2021-04-12 at 2.28.04 PM.png](https://files.reviewableusercontent.io/images/reviewable/3106437/oFhJrvMHd9NirMYW5on11OYa/Screen%20Shot%202021-04-12%20at%202.28.04%20PM.png)
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5CLZh0yTfQbrEd1Sj:-MY5CLZh0yTfQbrEd1Sk:b-wd8cdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L366)):*
> ```XML
>   );
> 
>   u: any;
> ```

nit. It can be typed as `u: uPlot`
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5BfbL0S_-p7ntAa7r:-MY5BfbM0P7b8ZXhHNXW:b-x6ydx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L367)):*
> ```XML
> 
>   u: any;
>   el: HTMLDivElement;
> ```

React API provides `React.createRef` function to directly interact with DOM elements.
```
el = React.createRef<HTMLDivElement>();
```
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5Am1oB5KUjLBr0Qha:-MY5Am1oB5KUjLBr0Qhb:b-mo7w41) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L391)):*
> ```XML
> 
>       const uPlotData = touPlot(data);
>       this.u = new uPlot(options, uPlotData, this.el);
> ```

It might be better to construct new instance of `uPlot` in `componentDidMount` lifecycle hook - so it'll be called once when DOM elements are accessible.
In this case `componentDidUpdate` hook can be responsible to update data only.

``` typescript
  u: uPlot;
  el = React.createRef<HTMLDivElement>();

  componentDidMount() {
    const data = this.props.data;
    const metrics = this.metrics(this.props);
    const axis = this.axis(this.props);
    const options = configureUPlotLineChart(
      metrics,
      axis,
      data,
      this.props.timeInfo,
    );
    const uPlotData = touPlot(this.props.data);
    this.u = new uPlot(options, uPlotData, this.el.current);
  }

  componentDidUpdate(prevProps: Readonly<LineGraphProps>) {
    if (prevProps.data != this.props.data) {
      this.u.setData(touPlot(this.props.data));
    }
  }
```
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5b4kPCQqfgYXPhKPw:-MY5b4kPCQqfgYXPhKPx:bnnbpv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L394)):*
> ```XML
>     }
>     if (this.u && prevProps.data != this.props.data) {
>       this.u.setData(touPlot(this.props.data));
> ```

in addition to updating data for chart, we also probably need to update chart options (which computation is based on `data`).
I didn't find any special method in `uPlot` API for this, so probably in this case it might be okay to recreate chart. @dhartunian , what do you think?
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5FKyX0sWTi3yJmFo_:-MY5FKyX0sWTi3yJmFoa:bgna8eg) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L417)):*
> ```XML
>             ref={(el) => {
>               this.el = el;
>             }}
> ```

in addition to comment about `this.el` initialization above, ref can be consumed like this:
``` javascript
<div ref={this.el} />
```


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Dismissed @nathanstilwell from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63087)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster, @koorosh, @nathanstilwell, and @nkodali)

___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgNJvu2U5s5xl16lx_:-MYG59Lt5jaFNY_RTS7O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L308)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

> ```
>   const xValues: number[] = [];
>   data.results[0].datapoints.forEach((datapoint) => {
>     xValues.push(NanoToSeconds(datapoint.timestamp_nanos.toNumber()));
>   });
> ```

This looks a lot like a `.map`.

```suggestion
const xValues: number[] = data.results[0].datapoints.map(d => 
    NanoToSeconds(d.timestamp_nanos.toNumber()))
```
</blockquote></details>

Done.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY4aPwb6-5_odD448nR:-MYG0oW9CW6h0D3Iz9na:bfk7fip) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L309)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@dhartunian , current data points transformation function relies on the shape of the first series and may work in case all series have the same length.
But it looks like series can have different length that might cause incorrect points mapping.
As an example, if one node is dead, Hardware metrics don't include any data points for such node
![Screen Shot 2021-04-12 at 2.28.04 PM.png](https://files.reviewableusercontent.io/images/reviewable/3106437/oFhJrvMHd9NirMYW5on11OYa/Screen%20Shot%202021-04-12%20at%202.28.04%20PM.png)

</blockquote></details>

Done. Thanks for pointing this out! Code has been revised to take a union of all timestamps across all result lists and then cross-references y-values to that set of timestamps. That should alleviate this issue.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgNwtkEB2S3lv8vGOQ:-MYG5D4v46jwAj4zln8r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L313)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

> ```
>   const yValues: (number | null)[][] = [];
>   data.results.forEach((result) => {
>     yValues.push(result.datapoints.map((d) => d.value));
>   });
> ```

... maybe for this one a `.reduce`?

```js
  const yValues: (number | null)[][] = 
  data.results.reduce(
    (values, result) => [...values, result.datapoints.map((d) => d.value)],
    [],
  );
```
</blockquote></details>

Done.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5CLZh0yTfQbrEd1Sj:-MYG5QCeBntw_15K-MyM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L366)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. It can be typed as `u: uPlot`
</blockquote></details>

Done.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5BfbL0S_-p7ntAa7r:-MYG5oLq6dGvpOMh7sAn:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L367)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

React API provides `React.createRef` function to directly interact with DOM elements.
```
el = React.createRef<HTMLDivElement>();
```
</blockquote></details>

done.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5Am1oB5KUjLBr0Qha:-MYG68lU-dMWJCIXkXtn:b-1l4l08) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L391)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

It might be better to construct new instance of `uPlot` in `componentDidMount` lifecycle hook - so it'll be called once when DOM elements are accessible.
In this case `componentDidUpdate` hook can be responsible to update data only.

``` typescript
  u: uPlot;
  el = React.createRef<HTMLDivElement>();

  componentDidMount() {
    const data = this.props.data;
    const metrics = this.metrics(this.props);
    const axis = this.axis(this.props);
    const options = configureUPlotLineChart(
      metrics,
      axis,
      data,
      this.props.timeInfo,
    );
    const uPlotData = touPlot(this.props.data);
    this.u = new uPlot(options, uPlotData, this.el.current);
  }

  componentDidUpdate(prevProps: Readonly<LineGraphProps>) {
    if (prevProps.data != this.props.data) {
      this.u.setData(touPlot(this.props.data));
    }
  }
```
</blockquote></details>

In `componentDidMount` I don't have access to any data since it's still loading. A lot of the configuration relies on the data being loaded so I ended up doing it in `componentDidUpdate`.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 393 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgUKHb3YRgadABSrn8:-MYG6PgFBUcYzcvCnipz:b-f92xw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L393)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

A couple comments. 

1. This is JavaScript so, "strict equality" please. :smile: (`!==` vs `!=`)
2. Isn't `data` an object? Are you expecting a _deep equals_ here? Or is this just to check that the props changed (in which case this works because it's a _different object_). **






** I haven't finished my coffee yet this morning, so I'm betting this is just a props change check. If so feel free to ignore. 
</blockquote></details>

1. Done.
2. Yes all I care is that it's a different object which means I need to recompute the chart. I could do a deep equals to see if the data is in fact the same but I feel like it's not worth the hassle since it seems to work fine at the moment.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5b4kPCQqfgYXPhKPw:-MYG6jqD2khXuWsILUHc:bwjhjre) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L394)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

in addition to updating data for chart, we also probably need to update chart options (which computation is based on `data`).
I didn't find any special method in `uPlot` API for this, so probably in this case it might be okay to recreate chart. @dhartunian , what do you think?
</blockquote></details>

I've updated a lot of the chart options configuration to be functions which get recomputed when data changes. I think that should take care of this concern.
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MY5FKyX0sWTi3yJmFo_:-MYG6qk24JjMnmsCg-Bw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L417)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

in addition to comment about `this.el` initialization above, ref can be consumed like this:
``` javascript
<div ref={this.el} />
```
</blockquote></details>

Done.
___
*[pkg/ui/src/views/cluster/util/graphs.ts, line 595 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MXgWVdDDOHZ_SUd30Vg:-MYG6xcxB9MHuXWsp88l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650da90c3ff862617b94095fcc647e843a2b19a7/pkg/ui/src/views/cluster/util/graphs.ts#L595)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

just pointing out a _dangling_ `a` on the end of this sentence
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dhartunian commented:
@leeoniya is there a way to dynamically update an axis label in uPlot? I didn't see a function type available there. I'd like to be able to change the label when the scale is adjusted (from "bytes (KiB)" to "bytes (MiB)" for instance). Happy to file a PR but just want to check first that I'm not missing an existing way to do it.

leeoniya commented:
mutate and call `redraw()`: https://jsfiddle.net/0k5m9afn/

also related: https://github.com/leeoniya/uPlot/issues/437

dhartunian commented:
@leeoniya Thanks! That worked. Looks like if I'm calling `setData` in that scenario I don't need to also call `redraw`.

dhartunian commented:
@koorosh can you take a look at this? I'd like to merge this week. One place where I need help is with the CSS styling on the hovering legend. The goal with this PR is to replace existing nvd3 functionality without changing the design/style too much. In the near future we'll make a better legend but I'd like to keep the hovering one for now but just make it work *okay* in the meantime.

My problem currently is this:
![Screen Shot 2021-04-26 at 4 53 37 PM](https://user-images.githubusercontent.com/986307/116149289-f41d9880-a6af-11eb-8257-bccd6f17be51.png)
I'd like to either implement "smart" placement for the legend using popper or a library like that, or make it display multi-column when needed. I fiddled with CSS a bit but couldn't get flexbox to behave well since my skills are not so good there. I'll send you the URL of a large cluster you can test with.

koorosh committed:
ui: metrics legend is under graph when large

Currently, most of the charts either display multiple series for every node
in the cluster or some aggregated data with less than 10 series on a chart.
The problem arises when cluster has large amount of nodes, so then legend on
a chart cannot fit all information on a screen.

As a solution, this change handles two possible cases when chart has:
- less than 10 series. It covers cases for aggregated data on a chart
or for clusters with small number of nodes. In this case, the behavior of
legend isn't changed, it sticks to mouse pointer and displays series
as a single column;
- more than 10 series, then legend is displayed right under the chart and
visually looks like it is extended chart's section. Series are placed inline
to fill entire width and minimize growth in height.

Release note (ui change): on the metrics page, place the legend under
chart if more than 10 series are being displayed

koorosh commented:
@dhartunian , here's my attempt to adjust legend for large number of series: https://github.com/dhartunian/cockroach/pull/3

The main idea is to stick legend to the bottom of the page only in case chart has more than 10 series. Otherwise it displays legend in tooltip without any changes.

#### more than 10 series
<img width="995" alt="Screen Shot 2021-04-27 at 3 38 42 PM" src="https://user-images.githubusercontent.com/3106437/116246937-13c9c500-a773-11eb-8a5c-ef8bedca1842.png">

#### less than 10 series
<img width="994" alt="Screen Shot 2021-04-27 at 3 38 50 PM" src="https://user-images.githubusercontent.com/3106437/116246798-f563c980-a772-11eb-9a00-c2a5878ad24a.png">

@Annebirzin , what do you think about such behavior?

dhartunian commented:
@koorosh I like that idea! so it would just show up on hover and we'd position absolute only in cases where the list is shorter. Feel free to push a commit on top of mine :)

koorosh commented:
> and we'd position absolute only in cases where the list is shorter

in fact, it will position absolute everywhere with only difference that for short legend - tooltip will follow the mouse pointer, and for larger - it will stick under chart and overlap elements below the chart without shifting charts below (thanks to absolute positioning).

> Feel free to push a commit on top of mine :)

cannot do this due to restricted permissions (for code maintainers only).

Annebirzin commented:
@koorosh nice, I think this is works. If possible it might be nice to set a fixed height to the legend when there are 10+ nodes (maybe only the top 3 or 4 rows of nodes show and then the user scrolls to see the rest)

dhartunian commented:
@Annebirzin if we do fixed height, we have to show it all the time instead of only on hover. Would you prefer that?

Annebirzin commented:
Ah sorry I didn’t realize this appeared on hover. Maybe it will make sense
to keep on hover, and then we can try to implement the scrollable right
panel in the follow up iteration?


On Tue, Apr 27, 2021 at 1:20 PM David Hartunian ***@***.***>
wrote:

> @Annebirzin <https://github.com/Annebirzin> if we do fixed height, we
> have to show it all the time instead of only on hover. Would you prefer
> that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/63087#issuecomment-827776364>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMMD6UL4EOYP3H3TGLVRWLDTK3W7DANCNFSM42M2JAPQ>
> .
>

dhartunian commented:
Thx for the PR @koorosh. I've added your commit here. Can you review the rest of my code? 

Additionally, is there anything we can do about the layout of the large legend grid? It looks a bit messy and misaligned because the width of the elements changes when data is inserted into them. I tried CSS grid briefly but didn't find an easy solution here. Tough to make the legend re-align everything when, for example, one of the values jumps from zero to 200 and takes up much more space.

<img width="1163" alt="Screen Shot 2021-04-28 at 12 10 48 AM" src="https://user-images.githubusercontent.com/986307/116346189-be63d700-a7b7-11eb-8111-36da2886a6c2.png">

koorosh reviewed (commented):
> Reviewed 1 of 7 files at r1, 6 of 10 files at r2, 4 of 4 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63087)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @elkmaster, @nathanstilwell, and @nkodali)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMairo1oB7U7lGIb1Q:-MZMairo1oB7U7lGIb1R:b-es347s) (no related file):*
@dhartunian , it looks good to me with minor comments regarding to imports and type definitions and one question for discussion about limiting min time rage for selected time range on a chart.

---
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMSsT0BQAcTtmup33x:-MZMSsT0BQAcTtmup33y:b6slyy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L43)):*
> ```XML
> import uPlot from "uplot";
> import "uplot/dist/uPlot.min.css";
> import { findClosestTimeScale } from "oss/src/redux/timewindow";
> ```

nit. remove `oss/` prefix from import path
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 408 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMUEH21aSqcYqzNAIQ:-MZMUEH21aSqcYqzNAIR:b-xbmlxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L408)):*
> ```XML
>   setNewTimeRange(startMillis: number, endMillis: number) {
>     const start = startMillis / 1e3;
>     const end = endMillis / 1e3;
> ```

nit. can be reused `MilliToSeconds` func from `convert.ts` defined above.
___
*[pkg/ui/src/views/cluster/util/graphs.ts, line 690 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMWvoa5WBL-u7Jn6qh:-MZMWvoa5WBL-u7Jn6qi:bc1qgj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3976d2e40ff365daf42e1098b34f9df6040e1f3/pkg/ui/src/views/cluster/util/graphs.ts#L690)):*
> ```XML
>           // of the user's selection. Then I use the `posToIdx` to tell me
>           // what the xAxis range is of the pixels.
>           setTimeRange(
> ```

@dhartunian  should we limit the min size of time window that allowed for selection? Currently user can select at least 10 min range from the dropdown. 
To make it consistent, the same restriction might be added here. 
___
*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMYV-DDKfdp6U82QII:-MZMYV-DDKfdp6U82QIJ:b-elxtl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L39)):*
> ```XML
> import TimeSeriesQueryDerivative = protos.cockroach.ts.tspb.TimeSeriesQueryDerivative;
> import Long from "long";
> import { History } from "@types/history";
> ```

correct import path for History type:
`import { History } from "history";`
___
*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMY8TTCazCVxzj-W_6:-MZMY8TTCazCVxzj-W_7:bih5jqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L41)):*
> ```XML
> import { History } from "@types/history";
> import { TimeWindow, TimeScale } from "oss/src/redux/timewindow";
> import { PayloadAction } from "oss/src/interfaces/action";
> ```

nit. remove `oss/` prefixes from import paths
___
*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMYdJ0BdGEeRdelCJ1:-MZMYdJ0BdGEeRdelCJ2:b-h72ub7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L164)):*
> ```XML
>   sources?: string[];
>   setTimeRange: (TimeWindow) => PayloadAction<TimeWindow>;
>   setTimeScale: (TimeScale) => PayloadAction<TimeScale>;
> ```

nit. correct function type definition (requires to indicate dummy argument variable):
`setTimeRange: (ts: TimeWindow) => PayloadAction<TimeWindow>;`
___
*[pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMZRdWERP_3MzSrsx_:-MZMZRdX3_wLdiPl7UMk:blkjv5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3976d2e40ff365daf42e1098b34f9df6040e1f3/pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx#L33)):*
> ```XML
> import { PayloadAction } from "oss/src/interfaces/action";
> import { TimeWindow, TimeScale } from "oss/src/redux/timewindow";
> import { History } from "@types/history";
> ```

nit. the same as above: correct import path for History type:
`import { History } from "history";`
and remove "oss" prefixes.
___
*[pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx, line 94 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMZmGE6jtgu7ji2Zvl:-MZMZmGE6jtgu7ji2Zvm:b-wvern7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3976d2e40ff365daf42e1098b34f9df6040e1f3/pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx#L94)):*
> ```XML
>   requestMetrics: typeof requestMetricsAction;
>   setTimeRange: (TimeWindow) => PayloadAction<TimeWindow>;
>   setTimeScale: (TimeScale) => PayloadAction<TimeScale>;
> ```

nit. the same as above. Correct function type definition (requires to indicate dummy argument variable):
`setTimeRange: (ts: TimeWindow) => PayloadAction<TimeWindow>;`


<!-- Sent from Reviewable.io -->

koorosh commented:
> Additionally, is there anything we can do about the layout of the large legend grid? It looks a bit messy and misaligned because the width of the elements changes when data is inserted into them. 

@dhartunian , as a quick solution I would propose to set `min-width` for legend rows to the value that's big enough to contain potentially any node labels without resizing elements.

Let's assume that longest legend's item looks like `(n9999) 255.255.255.255:65536 888888 GiB`
Chart's width and legend's font size are predefined as well.

Having this, it is possible to set `min-width: 30%` which in turn emulates 3 columns in legend and allows to increase the size of items if needed.

Required change for this (`pkg/ui/src/views/cluster/components/visualization/visualizations.styl`):
```styl
    > tr:first-child
      width 100%
      font-size $font-size--small

    > tr
      min-width 30%
```

<img width="997" alt="Screen Shot 2021-04-28 at 12 36 13 PM" src="https://user-images.githubusercontent.com/3106437/116384130-39f96e80-a820-11eb-9738-df9843ff98ff.png">


https://user-images.githubusercontent.com/3106437/116384167-4251a980-a820-11eb-8769-ea96f4889aba.mov

dhartunian commented:
@koorosh I'd like to merge https://github.com/cockroachdb/cockroach/pull/64367 prior to this one so we get TS issues out of the way.

dhartunian reviewed (commented):
@koorosh Nice. Thanks for the `min-width` tip. Added.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-:-MZOiwyd6S0RqX8r6PWe:b-uqq3ls)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster, @koorosh, @nathanstilwell, and @nkodali)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMairo1oB7U7lGIb1Q:-MZOwGR_7UmO3xCwxhPX:bhg7ya3) (no related file):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@dhartunian , it looks good to me with minor comments regarding to imports and type definitions and one question for discussion about limiting min time rage for selected time range on a chart.
</blockquote></details>

:thumbsup: TFTR I cleaned it all up.

---
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMSsT0BQAcTtmup33x:-MZOj3hM74_Mh1Og0z8G:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L43)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. remove `oss/` prefix from import path
</blockquote></details>

done
___
*[pkg/ui/src/views/cluster/components/linegraph/index.tsx, line 408 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMUEH21aSqcYqzNAIQ:-MZOjNls0KSbmUyAVg_B:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/cluster/components/linegraph/index.tsx#L408)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. can be reused `MilliToSeconds` func from `convert.ts` defined above.
</blockquote></details>

Done
___
*[pkg/ui/src/views/cluster/util/graphs.ts, line 690 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMWvoa5WBL-u7Jn6qh:-MZOjq_A15AZOHGZdBYv:b-r3v9ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3976d2e40ff365daf42e1098b34f9df6040e1f3/pkg/ui/src/views/cluster/util/graphs.ts#L690)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@dhartunian  should we limit the min size of time window that allowed for selection? Currently user can select at least 10 min range from the dropdown. 
To make it consistent, the same restriction might be added here. 
</blockquote></details>

Added enforcement of 10m minimum in the `setTimeRange` function definition in `linegraph/index.tsx`
___
*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMYV-DDKfdp6U82QII:-MZOk22cDBJMkTDblMAL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L39)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

correct import path for History type:
`import { History } from "history";`
</blockquote></details>

Done.
___
*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMY8TTCazCVxzj-W_6:-MZOk2R_4lF3XrSPce_2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L41)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. remove `oss/` prefixes from import paths
</blockquote></details>

Done.
___
*[pkg/ui/src/views/shared/components/metricQuery/index.tsx, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMYdJ0BdGEeRdelCJ1:-MZOw-xa6NUAHZapf75E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a12fa5cbb7cc26de997fb993b97f57fe3684fed9/pkg/ui/src/views/shared/components/metricQuery/index.tsx#L164)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. correct function type definition (requires to indicate dummy argument variable):
`setTimeRange: (ts: TimeWindow) => PayloadAction<TimeWindow>;`
</blockquote></details>

Done.
___
*[pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMZRdWERP_3MzSrsx_:-MZOwB9uBp30fAoHychL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3976d2e40ff365daf42e1098b34f9df6040e1f3/pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx#L33)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. the same as above: correct import path for History type:
`import { History } from "history";`
and remove "oss" prefixes.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx, line 94 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-MZMZmGE6jtgu7ji2Zvl:-MZOwAfFC-9QZWPpWokH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3976d2e40ff365daf42e1098b34f9df6040e1f3/pkg/ui/src/views/shared/containers/metricDataProvider/index.tsx#L94)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. the same as above. Correct function type definition (requires to indicate dummy argument variable):
`setTimeRange: (ts: TimeWindow) => PayloadAction<TimeWindow>;`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

koorosh reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 13 files at r5, 10 of 11 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63087#-:-MZRbuLoCxzA5q5BFZQp:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @elkmaster, @nathanstilwell, and @nkodali)



<!-- Sent from Reviewable.io -->

dhartunian commented:
bors r+
